### ISSUE TITLE
Discussion for possible `@angular/elements` public API changes

### ISSUE THREAD

**Issue by gkalpak** [angular/angular#21585]
Possible `@angular/elements` public API (breaking) changes for discussion:

- [ ] Rename `NgElementConstructor.is` to `NgElementConstructor.name`.
  <details>
    <summary>Background / Previous discussion</summary>
    Originally, we called this `selector`. @robwormald suggested renaming to `is` to match what Polymer does. `is` is special according to the spec when defining "customized built-in elements" (which we currently do not support, but might support in the future).

    @IgorMinar said:
    Shouldn't this be called `element-name`? You can't use other kind of selector anyway.

    @gkalpak said:
    ATM `selector` and `nodeName`/`element-name` mean pretty much the same thing. But if we add support for "customized built-in elements", then `is` will be more accurate. Or maybe `name` or `definedName`, which is what [the spec calls it](https://html.spec.whatwg.org/multipage/custom-elements.html#concept-custom-element-definition-name).

    @IgorMinar said:
    `name` would work for me.
  </details>

- [ ] Remove the `NgElementConstructor#upgrade()` method.
  <details>
    <summary>Background / Previous discussion</summary>
    This method can be used to upgrade any DOM element to an Angular Custom Element.
    AFAICT, the initial reasoning was to be able to support environments that do not natively support Custom Elements. But using polyfills instead would result in a much better (and uniform) experience.

    We could hide it from the public API, but keep the implementation in case it turns out to be useful.
  </details>

- [ ] Support using a component in Angular templates _and_ exporting it as a custom element (in the same app).
  The solution we had previously agreed on (for lack of a better alternative), was to prefix the selectors of Custom Elements. (So, if you define an Angular component with selector `foo-bar`, the Custom Element will be registered with selector `<prefix>-foo-bar`.)
  <details>
    <summary>Background / Previous discussion</summary>
    **Issue:**
    ATM, when an Angular component is registered as a Custom Element, using it inside an Angular template causes it to be instantiated twice: Once by Angular’s compiler and once by the browser (which recognizes it as a Custom Element and calls its `connectedCallback()`). Essentially, there is no way to tell whether an element is part of an Angular view and therefore should not be treated as a custom element.

    **Example:**
    I want to publish my `<fancy-button>` and `<fancy-datepicker>` Angular components as Custom Elements (part of the same `NgModule`). `<fancy-datepicker>` uses `<fancy-button>` inside its template.

    **Ideal solution:** _(currently not possible)_
    Be able to detect from inside the `NgElementConstructor` (or at least the `connectedCallback()`) that something is inside an Angular view (e.g. in an Angular component’s template) and do nothing (i.e. not instantiate a component instance).

    **Possible work-arounds:**
    Use different selectors. E.g. `<fancy-button>` inside Angular templates and `<gk-fancy-button>` when used as a Custom Element.
    - _Option 1_: We could add a new optional property to the `@Component` decorator (e.g. `customElementSelector: string`), which would default to the value of the `selector` property (useful when there is no intention to use the same component both as a Custom Element and inside Angular templates on the same page).
    - _Option 2_: We could automatically prefix all selectors with a pre-defined (configurable) prefix when registering as custom elements (e.g. in `NgElements#register()`). So, `<fancy-button>` would be (by default) registered as `'nge-fancy-button'` (where `nge` is a configurable prefix). This has the added benefit that essentially "name-spaces" the components and makes it easier to avoid conflicts between different Custom Element libraries. (Although, this might be a moot point, since it is already recommended to prefix your components with a custom prefix.)
    - _Option 3_: Have an annotation which would tell the compiler that this element is CE. This way the renderer could create the element with a marker which would inhibit the CE facade from doing anything.

    @mhevery said:
    The only way I can think of doing this is for Renderer to put an expando on the DOM elements. That way when elements facade wakes up it can check for the presence of the expando and short circuit itself. The expando would have some cost on the DOM creation.
    I think the easiest thing to do for now is to have a different prefix/suffix for the elements.

    @gkalpak said:
    Since the CE constructor is called before Angular instantiates the component, could we set a special property from the CE constructor that Angular could check and ignore the element?
    (Will that be compatible with AoT?)

    @mhevery said:
    I don't think that is something that we could do since it would mean that ALL Angular apps would be penalized with the check regardless if they are using `@angular/elements`. I think our solution will have to be such that the onus is on `@angular/elements` to do the check. It's the CE which has to be suppressed rather than the Angular component.
    So I think CE needs to be able to detect that Angular component is active and than become a noop. Not sure what is the best way to do that without penalizing everyone.
    One idea is that we change the renderer to put an expand on all elements which have a dash. The issue is that checking if `-` dash is present and than monkey patching is not free, especially if you have to do the check for each element. So maybe the check would have to be done at compile time and then at runtime we would have to pass an extra info to the renderer. (Or maybe we don't have to change the compiler but change the view engine to do the check once per view).
  </details>

- [ ] Remove the second signature of `registerAsCustomElements()`. 
  <details>
    <summary>Background / Previous discussion</summary>
    ATM, `registerAsCustomElements()` supports two signatures:

    ```ts
    function registerAsCustomElements(
        customElementComponents: Type[],
        bootstrapFn: () => Promise<NgModuleRef>): Promise<NgModuleRef>;
    ```

    ```ts
    function registerAsCustomElements(
        customElementComponents: Type[],
        platformRef: PlatformRef,
        moduleFactory: NgModuleFactory): Promise<NgModuleRef>;
    ```

    The second is just a shorthand for:
    
    ```ts
    registerAsCustomElements(
        customElementComponents,
        () => platformRef.bootstrapModuleFactory(moduleFactory))
    ```

    Unless there is any benefit to it (e.g. it is easier for the CLI to auto-rewrite for JiT/AoT, then I don't see any benefit in keeping both.

  </details>

**Comment by robwormald** [comment]
> NgElementConstructor.is to NgElementConstructor.name.
I guess? `is` is shorter, but don't think this really matters much. 

> Remove the NgElementConstructor#upgrade() method.
This is still useful in cases where the CE API isn't available, and (imho) useful even if you're not using CE directly (its WAY simpler than the ComponentFactoryResolver shenanigans...)

> Support using a component in Angular templates and exporting it as a custom element (in the same app).

I don't think renaming selectors is a good idea. You can easily work around this by simply not importing the NgModule where the components are declared into other NgModules (and turning on CUSTOM_ELEMENTS_SCHEMA) in the module that uses the CEs. 

What this means in practice is the ideal world is really one Angular Component per NgModule per Angular Element - this allows them to go through the compiler and get compiled into NgFactories, and work standalone without any global knowledge, with the tradeoff that it's more verbose. 

Longer term, a Custom Element (Angular or otherwise) *should* be type-checkable with vanilla `d.ts` files, since the public API of a Custom Element (props, attributes, events) can be expressed without additional metadata: see https://goo.gl/c8AVzS 

> Remove the second signature of registerAsCustomElements().

This is probably fine? Ideally, we'd drop the need for a module(Ref) at all, but i think we're a ways out on that.

**Comment by gkalpak** [comment]
>> NgElementConstructor.is to NgElementConstructor.name.

> I guess? `is` is shorter, but don't think this really matters much. Polymer folks are moving away from it. `selector`? `tag`? 

I don't think it matters much now, but will probably matter when/if we add support for customized built-in elements.
For example, if you want a Custom Element that extends `button`, you have to define it and use it like this:

```
customElements.define('fancy-button', ..., {extends: 'button'});
<button is="fancy-button"></button>
```

In that case, the element's _defined name_ (or _is_) will be `fancy-button`, the _tag_ will be `button` and the _selector_ will be `button[is="fancy-button"]` 😱 

So, I wouldn't use `selector` or `tag`. Or if we do, I would be carefull not to encourage patterns like `customElements.define(NgElementFactory.tag, NgElementFactory)`, because they may break in the future.

---
>> Support using a component in Angular templates and exporting it as a custom element (in the same app).

> I don't think renaming selectors is a good idea. You can easily work around this by simply not importing the NgModule where the components are declared into other NgModules (and turning on CUSTOM_ELEMENTS_SCHEMA) in the module that uses the CEs. 

This would mean that it is not possible to use a Custom Element as an Angular component. It doesn't sound like a good idea to me.

Would that also mean that the consumer is responsible for knowing what my component's dependencies are and include them manually? For example, in the `<fancy-calendar>`/`<fancy-button>` usecase mentioned in my original comment, if a consumer is only interested in my `<fancy-calendar>` Custom Element, they must still load my `<fancy-button>` Custom Element (which `<fancy-calendar>` depends on). Can we automate that?

(Maybe it's a reasonable first-version trade-off. I don't feel too strongly about it. Just pointing it out.)

> What this means in practice is the ideal world is really one Angular Component per NgModule per Angular Element - this allows them to go through the compiler and get compiled into NgFactories, and work standalone without any global knowledge, with the tradeoff that it's more verbose. 

Off the top of my head, another trade-off will be that they won't share the same `NgZone` instance (unless explicitly wired up like that). In contrast, all Custom Elements that come from a single `NgModule` will share the same `NgZone` instance.
